swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -727,6 +754,30 @@ predicate storeStep(Node node1, ContentSet c, Node node2) {
     c.isSingleton(any(Content::ArrayContent ac))
   )
   or
+  // read of a dictionary value via subscript operator, with intermediate step
+  exists(AssignExpr assign, SubscriptExpr subscript | 
+    subscript = assign.getDest() and
+    (
+      subscript.getArgument(0).getExpr() = node1.asExpr() and
+      node2.(DictionarySubscriptNode).getExpr() = subscript and
+      c.isSingleton(any(Content::TupleContent tc | tc.getIndex() = 1))
+      or
+      assign.getSource() = node1.asExpr() and
+      node2.(DictionarySubscriptNode).getExpr() = subscript and
+      c.isSingleton(any(Content::TupleContent tc | tc.getIndex() = 1))
+      or
+      node1.(DictionarySubscriptNode) = node1 and
+      node2.asExpr() = subscript and
+      c.isSingleton(any(Content::CollectionContent cc))
```

Comment:
So now we do:
```
source -> TupleContent -> CollectionContent
```
in two steps (which makes sense), and I gather that the last of these three cases is the `CollectionContent` step. But why do we need two cases to cover the first step? I would have thought that only the middle cases was needed, and the first case would happen through reverse reads?

Reply:
I think there are two steps to assemble the tuple because one is for the key and one is for the value.